runtime/src/bank.rs

```diff
@@ -6803,6 +6875,22 @@ impl Bank {
             }
         }
 
```

Comment:
This is probably incorrect because it makes the following assumptions:
- That only one `Bank` per validator is freshly initialized (not cloned from a parent).
- That the freshly initialized `Bank` has the lowest `Slot` that will be replayed, so that it generates the old environment when the validator is starting up on an `Epoch` boundary.

Reply:
In what scenario would the first assumption be incorrect?
I can see two code paths through which `finish_init()` could be called.
1. `process_blockstore_for_bank_0()` -> `Bank::new_with_paths()` -> `bank.finish_init()`
2. snapshots -> `reconstruct_bank_from_fields()` -> `Bank::new_from_fields()` -> `bank.finish_init()`

1st should only be called for bank0, i.e. for the first ever bank in the system. For 2nd, I believe it's called for the root bank if the validator is booted up from the snapshot.

Reply:
Yes, assumption 1 probably holds, but what about 2?

Reply:
I think 2 should also be ok. If the bank is rooted (or, is bank0), there should be no older slot that should get replayed. Wdyt?

Reply:
The important thing is that the oldest slot gets replayed first.